src/luis-api-client.ts

```diff
@@ -14,252 +14,496 @@
+        phrases: string;
+    }
+
+    export interface LabeledEntity {
+        name: string;
+        startToken: number;
+        endToken: number;
+        word: string;
+        isBuiltInExtractor: boolean;
     }
 
-    export interface UpdateUtteranceResult {
+    export interface LabeledUtterance {
+        id: string;
         utteranceText: string;
-        has_error: boolean;
-        error: string;
+        intent: string;
+        entities: LabeledEntity[];
     }
 
-    export interface TrainingStatusDetails {
-        statusId: number;
-        status: string;
-        trainingTime: string;
-        exampleCount: number;
-        trainingDuration: string;
-        failureReason: string;
+    export interface Entity {
+        entityType: string;
+        startToken: number;
+        endToken: number;
+    }
+
+    export interface Example {
+        exampleText: string;
+        selectedIntentName: string;
+        entityLabels?: Entity[];
     }
 
-    export interface TrainingStatus {
+    export enum TrainingStatuses { UpToDate = 2, InProgress = 3 }
+
+    export interface ModelTrainingStatus {
         modelId: string;
-        details: TrainingStatusDetails;
+        status: TrainingStatuses;
+        exampleCount: number;
+        failureReason?: string;
     }
 
-    export interface AppPublishData {
+    export type TrainingStatus = ModelTrainingStatus[];
+
+    export interface PublishResult {
         url: string;
-        previewUrl: string;
         subscriptionKey: string;
-        publishDate: string;
+        publishDate: Date;
     }
 }
 
-export class LuisClient {
-    BASE_API_PATH = 'https://api.projectoxford.ai/luis/v1.0/prog/apps/';
-    _baseRequest: any = null;
-    DEFAULT_PAGE_SIZE = 100;
+const RETRY_OPTS = {
+    retries: 10,
+    factor: 2,
+    minTimeout: 1500
+};
+// Maximum number of parallel requests to the API (to minimize 429 rate errors)
+const REQUESTS_PER_SECOND = 4;
+// count parameter of the getExamples API (the API doesn't support more than 100)
+const MAX_EXAMPLES_COUNT = 100;
+// Number of parallel getExamples requests to get all the examples (still rated by REQUESTS_PER_SECOND)
+const MAX_PARALLEL_EXAMPLES_REQUESTS = 15;
+// Maximum number of examples that can be created at the same time (the API doesn't support more than 100)
+const MAX_EXAMPLES_UPLOAD = 100;
 
-    constructor(subscriptionId: string) {
-        this._baseRequest = request.defaults({
-            baseUrl: this.BASE_API_PATH,
+export interface LuisApiClientConfig {
+    baseUrl?: string;
+    applicationId: string;
+    subscriptionKey: string;
+    requestsPerSecond?: number;
+}
+
+export class LuisApiClient {
+    protected applicationId: string = null;
+    private readonly req: any;
+    private readonly promiseThrottle: any;
+
+    constructor(config: LuisApiClientConfig) {
+        this.applicationId = config.applicationId;
+        let baseUrl = config.baseUrl || 'https://westus.api.cognitive.microsoft.com';
```

Comment:
Mmmm, it is not exactly the same because constants are values that a developer could change to tune algorithms, while that is a fixed value. Nevertheless, I'll put it on a constant, make sense.